File: DSA Practice/LeetCode/236. Lowest Common Ancestor of a Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode dfs(TreeNode root, TreeNode p, TreeNode q){
        if (root == p || root == null || root == q){
            return root;
        }

        TreeNode leftResult = dfs(root.left, p, q);
        TreeNode rightResult = dfs(root.right, p, q);

        if (leftResult == null && rightResult == null){
            return null;
        }

        if (leftResult != null && rightResult != null){
            return root;
        }

        return leftResult == null ? rightResult : leftResult;
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode res = dfs(root, p, q);
        return res;
    }
}

File: DSA Practice/LeetCode/236. Lowest Common Ancestor of a Binary Tree/solution.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun dfs(root: TreeNode?, p: TreeNode?, q: TreeNode?) : TreeNode? {
        if (root == p || root == q || root == null){
            return root
        }

        var leftResult : TreeNode? = dfs(root.left, p, q)
        var rightResult : TreeNode? = dfs(root.right, p, q)

        if (leftResult == null && rightResult == null){
            return null
        }

        if (leftResult != null && rightResult != null){
            return root
        }

        if (leftResult == null){
            return rightResult
        } else {
            return leftResult
        }
    }

    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        return dfs(root, p, q)
    }
}